File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    answer = 0
    gift = {}
    gidx = {}
    for f in friends:
        gift[f] = {}
        gidx[f] = 0
        
    
    for g in gifts:
        a,b = g.split(" ")
        if b in gift[a]:
            gift[a][b] +=1
        else:
            gift[a][b] = 1
        gidx[a] +=1
        gidx[b] -=1
    
    print(gift)
    
    get_gift = [0 for _ in friends]
    for i in range(len(friends)):
        friend = friends[i]
        for j in range(i+1,len(friends)):
            another = friends[j]
            a = gift[friend][another] if another in gift[friend] else 0
            b = gift[another][friend] if friend in gift[another] else 0
            
            if a > b:
                get_gift[i] +=1
            elif a<b:
                get_gift[j] +=1
            elif a ==b:
                ai,bi = gidx[friend], gidx[another]
                if ai>bi:
                    get_gift[i] +=1
                elif ai<bi:
                    get_gift[j] +=1
    answer = max(get_gift)
    
    return answer

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;
class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        int n= friends.length;
        
        HashMap<String,Integer> map =new HashMap<>();
        for(int i=0;i<n;i++){
            map.put(friends[i],i);
        }
        int[] giftDegree = new int[n];
        int[][] giftGraph = new int[n][n];
        
        
        for(String g:gifts){
            String[] gname = g.split(" ");
            giftDegree[map.get(gname[0])]++;
            giftDegree[map.get(gname[1])]--;
            giftGraph[map.get(gname[0])][map.get(gname[1])]++;
            
        }
        for(int i=0;i<n;i++){
            int cnt = 0;
            for(int j=0; j<n;j++){
                if(i==j)continue;
                if(giftGraph[i][j] > giftGraph[j][i] || 
                  (giftGraph[i][j]== giftGraph[j][i] && giftDegree[i]>giftDegree[j])) cnt++;
            }
            if(answer<cnt){
                answer= cnt;
            }
        }
        
        return answer;
    }
}